### ISSUE TITLE
Target Zones For Manual Placement Prompts

### ISSUE THREAD

**Issue by jeffersonrabb** [Automattic/newspack-popups#444]
### The Problem

The current model for placing prompts on the homepage or other one-off pages, is fairly awkward and problematic. The editor creates prompts with placement set to "manual placement," then uses the shortcode block to put them in the desired placement. There are at least three problems with this approach:

1) If there are multiple prompts targeted to different segments for the same spot, you would need to stack a bunch of shortcode blocks, which creates a pretty awkward editor UI.
2) To replace a prompts with a different one, the editor would have to manually find and replace every shortcode used on the site.
3) Manual Placement campaigns are generally (always?) going to be inline, so in the scenario where shortcode blocks are stacked there could be cases where more than one are shown to a user who matches multiple segments. 

### Proposed Solution

- Introduce the idea of prompt "zones" or "slots."
- New Gutenberg block allows a defined zone to be placed anywhere blocks are found. (Maybe zone definition actually occurs in the block?)
- Individual prompts placement UI will include the names of all zones. 
- Prompts whose placement is a zone will be displayed there according to normal rules of segmentation and precedence. 
- Much like overlays and above header placements, there can only be one prompt shown per zone. 

### Open Questions

- Should the block actually show the contents of the selected prompt, or should it just be the name of the zone? And... if the block does show the selected prompt and there are multiple options, which one should be shown?
- Where/how are zones defined?
- What is this thing called? Zone? Slot? Other?

cc: @brianboyer @dkoo @adekbadek @thomasguillot

**Comment by adekbadek** [comment]
I like the proposed solution. Basically we allow the user to define single-prompt-only positions, on top of overlay and above header.

> Should the block actually show the contents of the selected prompt, or should it just be the name of the zone? And... if the block does show the selected prompt and there are multiple options, which one should be shown?

If there is just one prompt assigned to the zone, it can be displayed, along with an edit link. If more than one, a list of prompt titles can be shown instead.

> Where/how are zones defined?

I like the proposed solution - the user is placing the zones as blocks, wherever they please. A zone has a unique ID and a human-friendly label assigned by the user.

> What is this thing called? Zone? Slot? Other?

I'm for slots.

**Comment by dkoo** [comment]
I think this is a feature/UI that could benefit from some user stories and basic wireframes before we start building it out. cc @thomasguillot if you have ideas on this!

Here's my attempt at writing up a user story and describing how it might be done with a new UI. Writing it up like this helps me understand what we're trying to build and see what questions we still need to answer. Happy to workshop this with the team. 

Note that this only really works after implementing #432, which will ensure that readers will only ever see a single prompt in each zone.

In the Campaigns dashboard, we visit a new **Zones** tab (or whatever terminology we end up choosing—going with "zones" as the terminology since it's the first term mentioned in the thread, but I don't have a strong preference). We create a single zone called "Donate or Signup" that will create a new option in the Placement dropdown, only for inline prompts. Right now the zone shows that it's empty (no prompts assigned) and has a popover menu that lets you delete or rename it. Deleting a zone with assigned prompts will reset those prompts' placement to "in article content" (the default placement for inline prompts). **Note:** I think it makes more sense to manage zones in the Campaigns dashboard rather than the block itself, since I'd imagine you'd want to create your zones and assign prompts to them before you're ready to place them in content. We could still offer links to create new zones from the block.

We create two inline prompts for manual placement. (When presented with the New Prompt modal, there is no longer a Manual Placement option, since all manual placement prompts are inline.)

* Donate CTA is assigned to the Newsletter Subscribers segment, which has top priority. We want to show this to readers who have already subscribed.
* Signup CTA is assigned to the New Readers segment, which has lower priority. We want to show this in the same spot, but to readers who haven't yet subscribed.

In the editor sidebar for both prompts, we choose the "Donate or Signup" zone from the Placement dropdown. Any zones should appear together in the dropdown under a "Manual Placement Zones" optgroup header. **Question:** Is there ever a scenario where we'd want to be able to assign a single prompt to more than one zone? If so, then I would propose that we separate the UI from the Placement control so that you'd select "Manual Placement" as the placement, and then show a multi-select control to select one or more zones only if that placement option is selected. This would also make things more complicated for showing zones in the prompt action cards in the dashboard UI, but we can defer until we decide whether multiple zones per prompt is a possibility.

We publish both prompts. At this point, despite being active, neither prompt will be ever be seen on the front-end because they're assigned to a manual placement zone which hasn't yet been added anywhere on the site.

In the Campaigns tab of the dashboard, the action cards for any inline prompts should have a new dropdown under the cog menu for placement, so that we can quickly assign zones without editing each individual prompt.

In the Zones tab, we now see action cards (or simplified title links) for the two prompts underneath the Donate or Signup zone heading. There is a potential conflict here if more than one prompt with the same segment is assigned to the same zone. In this case, we should treat conflicts like do for overlay and above-header prompts in the Campaigns tab: show a warning in the dashboard for conflicting prompts, and only show the most recently created prompt on the front-end. These warnings should also be shown with each conflicting prompt in the Campaigns tab.

We edit the homepage (plus whatever other posts or pages we want the prompts to appear in) and insert a **Newspack Campaigns Zone** block where we want the prompts to appear. The block is empty at first and allows selection from a dropdown (in the editor sidebar) of a single zone. There is also a button to open the Campaigns dashboard's Zones tab in a new window. If published without a zone selection, the block will render nothing on the front-end.

We select the "Donate or Signup" zone from the dropdown. The editor render changes to show the name of the selected zone as well as a list of the prompts assigned to that zone, with the zone name linking back to the Zones tab, and each prompt name linking back to the prompt's edit page. Perhaps this is even a partial recreation of the action card UI in the Zones tab, complete with conflict warnings. **Note:** I couldn't think of a way to render a preview of the actual prompts in this user story, in which we want to potentially show different prompts to different segments in the same zone, so showing some kind of list makes more sense to me. We update the page.

The block with a zone selection has a server-side render which renders `[newspack-popup]` shortcodes for each prompt in the zone, inside a container element. With non-cascading segment matching (#432), readers will only see the single prompt with the highest-priority segment they match (and if that segment isn't assigned to any prompts in the zone, they see nothing). If multiple prompts in the same zone have the same segment and this is the reader's highest-priority matching segment, then the conflict-resolution logic will only show the most recently created prompt.

**Comment by thomasguillot** [comment]
Slots

**Comment by thomasguillot** [comment]
> Should the block actually show the contents of the selected prompt, or should it just be the name of the zone?

For simplicity, I'd say a `Placeholder` with the name of the slot

**Comment by dkoo** [comment]
### Summary of discussion

Terminology: "slots" or "custom placements"

Approach (basically the same as above, but with predefined custom placements instead of allowing users to define their own):

* A **Custom Placements** tab with 2-3 predefined options that can be renamed (but not deleted or added to)
* When editing inline prompts, the custom placements appear together as options in the Placement dropdown
* The cog popover menu in prompt action cards should include Placement
* A new custom block that can be inserted in any page or post content (or widget?) with a very simple UI that lets you choose a single custom placement
* Once a custom placement is chosen for the block, it displays a simple list of prompts with that placement in the editor and preview
* Priority logic, conflict warnings and conflict resolution as described above (there can only be one inline prompt per segment shown per block placement, the prompt with the highest-priority matching segment gets shown, and if there is more than one prompt with the same segment and the same custom placement, only the most recently created prompt gets shown)

**Comment by jeffersonrabb** [comment]
> A Custom Placements tab with 2-3 predefined options that can be renamed (but not deleted or added to)

I think this could be considered a nice-to-have. If this adds significant complexity to the project I'd be comfortable with a V1 where the predefined options cannot be renamed and there is no tab. The custom placements will appear everywhere Placements are listed, which should be adequate for making the feature discoverable.

**Comment by dkoo** [comment]
One thing we're missing from [this plan](https://github.com/Automattic/newspack-popups/issues/444#issuecomment-780707531) is a migration strategy for existing prompts with the "Manual Placement" frequency. Since custom placements are replacing this frequency option, how should we handle those existing prompts? Do we automatically assign them to a default custom placement?

**Comment by adekbadek** [comment]
This migration strategy sounds good 👍

**Comment by jeffersonrabb** [comment]
One thought about migration strategy - the manual placement approach is relatively new and complicated to use, which probably means it hasn't been used much. We could throw together a script to determine which Newspack sites have any Manual Placement prompts. If the number is low (I suspect/hope it will be < 5) it might be more efficient to simply handle these by hand. What do you think?

**Comment by dkoo** [comment]
I think this is mostly what's needed for migrating the manual prompts programmatically: https://github.com/Automattic/newspack-popups/pull/467/commits/e50f966542772ba20a3778d41c2a1ddf1e23ffd4

If we also ensure that prompts with custom placements are still rendered via shortcode, it should be fairly seamless for those existing prompts.
